### PR TITLE
client/asset/eth: add RPC client

### PR DESCRIPTION
**Pull request by buck54321**
```
client/asset/eth:
Adds an RPC client that can maintain one or more websocket (preferred), http, or
ipc connections. Implements ethFetcher as well as bind.ContractBackend. Websocket
connections have a block header feed, so need to make fewer requests. For http
connections, the block header requests are metered to one per 10 seconds.
If no nonces are involved, the provider to use for a request is picked randomly,
but closely grouped requests involving nonces will attempt to use the same provider.

client/asset: Add Repeatable bool field to ConfigOption to allow multiple
inputs.

ui: Work with new repeatable input type.
```

### CODE REVIEW

File: client/asset/eth/multirpc.go [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+	m.lastProvider.stamp = time.Now()
+	m.lastProvider.Unlock()
+	return nil
+}
+
+func (m *multiRPCClient) sendTransaction(ctx context.Context, txOpts *bind.TransactOpts, to common.Address, data []byte) (*types.Transaction, error) {
+	nonce, err := m.nextNonce(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (chappjc) [inline client/asset/eth/multirpc.go:571]: This is based on code that's getting fixed in #1638 .  The nonce comes from txOps.Nonce, so shouldn't call `nextNonce` in here.
https://github.com/decred/dcrdex/pull/1638/files#diff-2603858a9b5db2a5039633341a2fd13fd7933ee616b3182cb83f32e6f3610cd3R228

File: client/asset/eth/multirpc.go [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+		maxFeeRate = new(big.Int).Mul(baseFees, big.NewInt(2))
+	}
+
+	txOpts := newTxOpts(ctx, m.creds.addr, val, maxGas, maxFeeRate, gasTipCap)
+
+	nonce, err := m.nextNonce(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting nonce: %v", err)
+	}
+	txOpts.Nonce = new(big.Int).SetUint64(nonce)
```
Comment (chappjc) [inline client/asset/eth/multirpc.go:677]: Yeah, so this is good.  Caller has the nonce mutex locked through txOpts -> send

**Review by martonp — COMMENTED** [review]
Looks really nice, but I was unable to get it to work on simnet. I used the ipc file to add a provider, and at first it says it has connected to a peer, but when I try to make a trade, it fails saying that it has no peers.

Also, there is a UI issue. If I have a native wallet already created, I cannot see the Recieve/Send buttons, and I cannot switch to the rpc wallet. When I have no wallets created, I do not see the option of creating a native wallet.

**Comment by buck54321** [comment]
> Looks really nice, but I was unable to get it to work on simnet. I used the ipc file to add a provider, and at first it says it has connected to a peer, but when I try to make a trade, it fails saying that it has no peers.

I think I've resolved some things now @martonp, if you're up for another spin. 

> Also, there is a UI issue. If I have a native wallet already created, I cannot see the Recieve/Send buttons, and I cannot switch to the rpc wallet. When I have no wallets created, I do not see the option of creating a native wallet.

Oh. Interesting. Cuz I've removed support for native wallets. I'll figure this out.

File: client/asset/eth/eth.go [hunk 3]
```diff
@@ -348,18 +376,18 @@ func genWalletSeed(entropy []byte) ([]byte, error) {
 
 // CreateWallet creates a new internal ETH wallet and stores the private key
 // derived from the wallet seed.
-func CreateWallet(createWalletParams *asset.CreateWalletParams) error {
-	if createWalletParams.Type != walletTypeGeth {
+func CreateWallet(cfg *asset.CreateWalletParams) error {
+	return createWallet(cfg, false)
+}
+
+func createWallet(createWalletParams *asset.CreateWalletParams, skipConnect bool) error {
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/eth.go:464]: In some live tests.

File: client/asset/eth/multirpc.go [hunk 4]
```diff
@@ -0,0 +1,1143 @@
...
+
+}
+
+func (m *multiRPCClient) sendSignedTransaction(ctx context.Context, tx *types.Transaction) error {
+	var lastProvider *provider
+	if err := m.withPreferred(func(p *provider) error {
+		lastProvider = p
+		m.log.Tracef("Sending signed tx via %q", p.host)
+		err := p.ec.SendTransaction(ctx, tx)
+		if err != nil && strings.Contains(err.Error(), core.ErrAlreadyKnown.Error()) {
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go]: Coming through the `(rpc.Client).CallContext`, we lose the type.

File: client/asset/eth/multirpc.go [hunk 5]
```diff
@@ -0,0 +1,1270 @@
...
+}
+
+func (m *multiRPCClient) getTransaction(ctx context.Context, txHash common.Hash) (tx *types.Transaction, h int64, err error) {
+	return tx, h, m.withPreferred(func(p *provider) error {
+		resp, err := getRPCTransaction(ctx, p, txHash)
+		if err != nil {
+			return err
+		}
+		tx = resp.tx
+		if resp.BlockNumber != nil {
```
Comment (martonp) [inline client/asset/eth/multirpc.go:598]: This `resp.BlockNumber` string has a "0x" in the front, causing the `hex.DecodeString` call to return an error. Also this function should return -1 for the height if the tx has not yet been mined.

**Review by martonp — COMMENTED** [review]
All the functionality seems to be working well now, except for the redemption confirmation which works after fixing `getTransaction`.

**Review by JoeGruffins — COMMENTED** [review]
partial review

**Comment by JoeGruffins** [comment]
Maybe unrelated, but seeing for the first time on loadbot:
```
2022-09-13 15:54:52.160 [INF] MANTLE:CMPD:STACKER:1: created wallet dcr:iJt1 on node alpha
2022-09-13 15:54:52.189 [ERR] LOADBOT: exec error (dcr) "./alpha": exit status 1: -4: wallet.SendOutputs: double spend:: wallet.processTransactionRecord: 2f3cce875cae0b95a5d9b4deb30fc578f64ca2d4183eea84b0b8271f3dc66572 conflicts with 8fe8d127a39ecccf0d902fb4cabfbb32b8913948f801aacf64e71d4a9c1ba179 by double spending 6d58f2b8f0fb8412a4405dd9a99938733cec08256593c9992ca9d1199a1a54c6:2

2022-09-13 15:54:52.189 [CRT] MANTLE:CMPD:SNIPER:0: exit status 1
2022-09-13 15:54:52.189 [ERR] LOADBOT: exec error (dcr) "./alpha": exit status 1: -4: wallet.SendOutputs: double spend:: wallet.processTransactionRecord: 2dd5e335aac0763164798611d737e2bfdcd6bbc23b614b368306e7d1cd1c305d conflicts with 8fe8d127a39ecccf0d902fb4cabfbb32b8913948f801aacf64e71d4a9c1ba179 by double spending ecbe09281f2ae1d0f1144c07c518b156941cc6e437b5badadbfae9b6788840fa:0

2022-09-13 15:54:52.189 [CRT] MANTLE:CMPD:PINGPONG:0: exit status 1
```
Seems to have been random as has not happened again.

**Comment by JoeGruffins** [comment]
Trying to set up with the alpha rpc, I guess home symbol is not expanded here and I get the error `2022-09-13 16:13:05.981 [ERR] CORE[CREATE]: error creating http client for `2022-09-13 16:23:48.491 [ERR] CORE[CREATE]: error creating http client for "~/dextest/eth/alpha/node/geth.ipc": dial unix ~/dextest/eth/alpha/node/geth.ipc: connect: no such file or directory` which is fine but it won't let me try again until I restart the app or wait for a panic that seems to come:
image

I can make infinite providers but cannot remove the boxes. Doesn't need to be fixed now I guess:
image

Working well using the simnet harness alpha node! My testnet has almost synced... probably. Will try to set up for testing there.

**Comment by JoeGruffins** [comment]
Loading up testnet, I am unable to change my wallet in settings for eth, there are no options:

image

File: client/asset/eth/multirpc.go [hunk 6]
```diff
@@ -0,0 +1,1274 @@
...
+	p.tip.failCount = 0
+	p.tip.Unlock()
+}
+
+// cachedTip retrieves the last known best header.
+func (p *provider) cachedTip() *types.Header {
+	stale := time.Second * 10
+	if p.ws {
+		stale = time.Minute * 2
+	}
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:106]: Right. We expect them to come through the subscription.

File: client/asset/eth/multirpc.go [hunk 7]
```diff
@@ -0,0 +1,1274 @@
...
+
+	var connections int
+	for _, p := range m.providerList() {
+		if _, err := p.bestHeader(ctx, m.log); err != nil {
+			m.log.Errorf("Failed to synchrnoize header from %s: %v", p.host, err)
+		} else {
+			connections++
+		}
+	}
+	// TODO: Require at least two if all connections are non-local.
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:460]: I'm not considering this anymore. Removing TODO.

File: client/asset/eth/multirpc.go [hunk 8]
```diff
@@ -0,0 +1,1274 @@
...
+
+func getRPCTransaction(ctx context.Context, p *provider, txHash common.Hash) (*rpcTransaction, error) {
+	var resp *rpcTransaction
+	err := p.ec.rpc.CallContext(ctx, &resp, "eth_getTransactionByHash", txHash)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, asset.CoinNotFoundError
+	}
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:582]: "geth" is pretty ambigious. You probably mean the `ethereum.NotFound` error, which is an `ethclient.Client` interpretation of the `nil` `(*TransactionAPI).GetTransactionByHash` result. But we're using `rpc.Client` directly, since `ethclient.Client` strips out the block and sender details.

```Go
func (ec *Client) TransactionByHash(ctx context.Context, hash common.Hash) (tx *types.Transaction, isPending bool, err error) {
	var json *rpcTransaction
	err = ec.c.CallContext(ctx, &json, "eth_getTransactionByHash", hash)
	if err != nil {
		return nil, false, err
	} else if json == nil {
		return nil, false, ethereum.NotFound
	} else if ...
}
```

File: client/asset/eth/multirpc.go [hunk 9]
```diff
@@ -0,0 +1,1274 @@
...
+			s = m
+		case error:
+			if errors.Is(err, m) {
+				return true
+			}
+			s = m.Error()
+		}
+		if strings.Contains(errStr, s) {
+			return true
+		}
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:661]: That's true, and I think I'm okay with it. I would understand if anyone was adamant that I do it different though.

File: client/asset/eth/multirpc.go [hunk 10]
```diff
@@ -0,0 +1,1274 @@
...
+func (m *multiRPCClient) withOne(providers []*provider, f func(*provider) error, acceptabilityFilters ...acceptabilityFilter) (superError error) {
+	readyProviders := make([]*provider, 0, len(providers))
+	for _, p := range providers {
+		if !p.failed() {
+			readyProviders = append(readyProviders, p)
+		}
+	}
+	if len(readyProviders) == 0 {
+		// Just try them all.
+		m.log.Tracef("all providers in a failed state, so acting like none are")
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:677]: I think this would be somewhat common when using a single provider.

File: client/asset/eth/multirpc.go [hunk 11]
```diff
@@ -0,0 +1,1274 @@
...
+	// Check for an unexpired cached header first.
+	var bestHeader *types.Header
+	for _, p := range m.providerList() {
+		h := p.cachedTip()
+		if h == nil {
+			continue
+		}
+		// This block choosing algo is probably too rudimentary. Really need
+		// shnuld traverse parents to a common block and sum up gas (including
+		// uncles?), I think.
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:779]: Isn't GasUsed the rough equivalent of work in ethereum? I'm not clear on how that works I guess.

File: client/asset/eth/multirpc.go [hunk 12]
```diff
@@ -0,0 +1,1274 @@
...
+			discard, propagate, fail := f(err)
+			if discard {
+				return nil
+			}
+			if propagate {
+				return err
+			}
+			if fail {
+				p.setFailed()
+			}
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:700]: I had that messed up. Thanks.

File: client/asset/eth/multirpc.go [hunk 13]
```diff
@@ -0,0 +1,1274 @@
...
+func (m *multiRPCClient) transactionConfirmations(ctx context.Context, txHash common.Hash) (confs uint32, err error) {
+	var r *types.Receipt
+	var tip *types.Header
+	var notFound bool
+	if err := m.withPreferred(func(p *provider) error {
+		r, err = p.ec.TransactionReceipt(ctx, txHash)
+		if err != nil {
+			if strings.Contains(err.Error(), "not found") {
+				notFound = true
+			}
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:903]: Can't use it here, because a subsequent provider might have it.

File: client/asset/eth/multirpc.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/asset/eth/multirpc.go:779]: I'm not either. It may be that more gas used has more weight.

[the remark this answers is not in the document]
Comment (chappjc) [reply client/asset/eth/multirpc.go:779]: I wouldn't guess that gas used has any thing to do with it either.  Not saying it doesn't, but are there any references on this?

File: client/core/core.go [hunk 14]
```diff
@@ -2317,6 +2317,10 @@ func (c *Core) loadWallet(dbWallet *db.Wallet) (*xcWallet, error) {
 		})
 	}
 	if err != nil {
+		if errors.Is(err, asset.ErrWalletTypeDisabled) {
+			subject, details := c.formatDetails(TopicWalletTypeDeprecated, unbip(assetID))
+			c.notify(newWalletConfigNote(TopicWalletTypeDeprecated, subject, details, db.WarningLevel, nil))
+		}
 		return nil, fmt.Errorf("error opening wallet: %w", err)
 	}
```
Comment (buck54321) [inline client/core/core.go:2325]: I started into a scheme where we just disable the wallet, but it was a can of worms. Instead, we're refusing to open a deprecated wallet in e.g. `eth.NewWallet` and issuing a warning. This is no problem for us right now, because the only other option for Ethereum is also a seeded wallet and will derive the same key. But if there was a need to export a seed from a deprecated wallet, we would need to get more sophisticated here.

**Comment by JoeGruffins** [comment]
panic while testing on testnet using infura:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xbd4e81]

goroutine 5948022 [running]:
decred.org/dcrdex/client/asset/eth.(*multiRPCClient).transactionReceipt(0xc0002462c0, {0x1f0c358, 0xc001212200}, {0xe0, 0x40, 0x25, 0xcc, 0xc7, 0xe5, 0xc7, ...})
        /home/joe/git/dcrdex/client/asset/eth/multirpc.go:536 +0x1c1
decred.org/dcrdex/client/asset/eth.(*assetWallet).confirmRedemption(0xc000246210, {0xc01124eae0?, 0x4110c7?, 0x20?}, 0xc011c40300, 0xc000dfe501?)
        /home/joe/git/dcrdex/client/asset/eth/eth.go:3237 +0xd8a
decred.org/dcrdex/client/asset/eth.(*ETHWallet).ConfirmRedemption(0x1f03d40?, {0xc01124eae0?, 0xc000dfe570?, 0x1?}, 0xc004954c90?)
        /home/joe/git/dcrdex/client/asset/eth/eth.go:3003 +0x25
decred.org/dcrdex/client/core.(*trackedTrade).confirmRedemption(0xc0006b6300, 0xc001031a00)
        /home/joe/git/dcrdex/client/core/trade.go:2538 +0x165
decred.org/dcrdex/client/core.(*Core).tick(0xc00016d680, 0xc0006b6300)
        /home/joe/git/dcrdex/client/core/trade.go:1777 +0x1c7d
decred.org/dcrdex/client/core.(*Core).tickAsset.func1()
        /home/joe/git/dcrdex/client/core/core.go:963 +0x45
created by decred.org/dcrdex/client/core.(*Core).tickAsset
        /home/joe/git/dcrdex/client/core/core.go:962 +0x277
```

**Comment by JoeGruffins** [comment]
Also, weird critical error for the server randomly:
```
2022-09-21 14:22:56.000 [CRT] MKT: Archivist failing. Last unexpected error: pq: column "preimage" is of type bytea but expression is of type integer
2022-09-21 14:22:56.000 [CRT] MKT: aborting epoch processing on account of failing DB: pq: column "preimage" is of type bytea but expression is of type integer
2022-09-21 14:22:56.000 [DBG] MKT: epoch pump drained for market eth_dcr
2022-09-21 14:22:56.000 [INF] MKT: Market "eth_dcr" stopped.
2022-09-21 14:22:56.000 [INF] MKT: Market "eth_dcr" suspended after epoch 103983611, persist book = true.
```

**Comment by buck54321** [comment]
> panic while testing on testnet using infura:
> 
> ```
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xbd4e81]
> 
> goroutine 5948022 [running]:
> decred.org/dcrdex/client/asset/eth.(*multiRPCClient).transactionReceipt(0xc0002462c0, {0x1f0c358, 0xc001212200}, {0xe0, 0x40, 0x25, 0xcc, 0xc7, 0xe5, 0xc7, ...})
>         /home/joe/git/dcrdex/client/asset/eth/multirpc.go:536 +0x1c1
> decred.org/dcrdex/client/asset/eth.(*assetWallet).confirmRedemption(0xc000246210, {0xc01124eae0?, 0x4110c7?, 0x20?}, 0xc011c40300, 0xc000dfe501?)
>         /home/joe/git/dcrdex/client/asset/eth/eth.go:3237 +0xd8a
> decred.org/dcrdex/client/asset/eth.(*ETHWallet).ConfirmRedemption(0x1f03d40?, {0xc01124eae0?, 0xc000dfe570?, 0x1?}, 0xc004954c90?)
>         /home/joe/git/dcrdex/client/asset/eth/eth.go:3003 +0x25
> decred.org/dcrdex/client/core.(*trackedTrade).confirmRedemption(0xc0006b6300, 0xc001031a00)
>         /home/joe/git/dcrdex/client/core/trade.go:2538 +0x165
> decred.org/dcrdex/client/core.(*Core).tick(0xc00016d680, 0xc0006b6300)
>         /home/joe/git/dcrdex/client/core/trade.go:1777 +0x1c7d
> decred.org/dcrdex/client/core.(*Core).tickAsset.func1()
>         /home/joe/git/dcrdex/client/core/core.go:963 +0x45
> created by decred.org/dcrdex/client/core.(*Core).tickAsset
>         /home/joe/git/dcrdex/client/core/core.go:962 +0x277
> ```
> 
> Unable to recover, keeps panicking when I start up.
> 
> Other than that working well!

My bad. Fixed, and I've actually tested it on testnet this time. I've got some order up on testnet dcr-eth tonight if you get back around to testing.

**Comment by buck54321** [comment]
> Also, weird critical error for the server randomly @chappjc:
> 
> ```
> 2022-09-21 14:22:56.000 [CRT] MKT: Archivist failing. Last unexpected error: pq: column "preimage" is of type bytea but expression is of type integer
> 2022-09-21 14:22:56.000 [CRT] MKT: aborting epoch processing on account of failing DB: pq: column "preimage" is of type bytea but expression is of type integer
> 2022-09-21 14:22:56.000 [DBG] MKT: epoch pump drained for market eth_dcr
> 2022-09-21 14:22:56.000 [INF] MKT: Market "eth_dcr" stopped.
> 2022-09-21 14:22:56.000 [INF] MKT: Market "eth_dcr" suspended after epoch 103983611, persist book = true.
> ```
> 
> full logs: [dcrdextestnet.txt](https://github.com/decred/dcrdex/files/9613245/dcrdextestnet.txt)

Uh. This is kinda nuts.

**Comment by JoeGruffins** [comment]
Trades worked!
Unrelated to this pr, but is this right? I try to market buy with 2 eth but that becomes .2 eth:
image
image

File: client/asset/eth/multirpc.go [hunk 15]
```diff
@@ -0,0 +1,1294 @@
...
+			if !ok {
+				// Subscription cancelled
+				return
+			}
+			log.Errorf("%q header subscription error: %v", p.host, err)
+			log.Info("Attempting to resubscribe to %q block headers", p.host)
+			sub, err = newSub()
+			if err != nil { // context cancelled
+				return
+			}
```
Comment (JoeGruffins) [inline client/asset/eth/multirpc.go:243]: Trying the nodeclient tests on testnet I will see:
```
2022-09-23 15:11:34.532 [ERR] ETHTEST: "goerli.infura.io" header subscription error: <nil>
2022-09-23 15:11:34.532 [WRN] ETHTEST: can't resubscribe to "goerli.infura.io" headers: client is closed
2022-09-23 15:11:36.135 [ERR] ETHTEST: "goerli.infura.io" header subscription error: <nil>
2022-09-23 15:11:36.135 [WRN] ETHTEST: can't resubscribe to "goerli.infura.io" headers: client is closed
```
Expected?

File: client/asset/eth/nodeclient_harness_test.go [hunk 16]
```diff
@@ -129,6 +122,21 @@ var (
 	//
 	// TODO: Make this also work for token tests.
 	isTestnet = false
```
Comment (JoeGruffins) [inline client/asset/eth/nodeclient_harness_test.go:124]: I'm running like `TARGS=rpc\ testnet go test --tags lgpl,harness` is this correct? It works but I'm not sure if the intended way to set `TARGS`

**Review by JoeGruffins — APPROVED** [review]

File: client/asset/eth/multirpc.go [hunk 17]
```diff
@@ -0,0 +1,1298 @@
...
+	n := len(parts)
+	if n <= 2 {
+		return host
+	}
+	return parts[n-2] + "." + parts[n-1]
+}
+
+func checkProvidersCompliance(ctx context.Context, walletDir string, providers []*provider, log dex.Logger) error {
+	var compliantProviders map[string]bool
+	path := filepath.Join(walletDir, "compliant-providers.json")
```
Comment (chappjc) [inline client/asset/eth/multirpc.go:1253]: Can you explain the purpose of "compliant-providers.json" somewhere?
This is for providers that are not known in the hard coded lists, and thus go through a compliance check during wallet bring up.  And if `checkProvidersCompliance` finds it in the json file, that means it has already passed the test and to skip running the compatibility tests each time?  Will you add a doc somewhere to that effect?

File: client/asset/eth/multirpc.go [hunk 18]
```diff
@@ -0,0 +1,1298 @@
...
+	}
+
+	log.Tracef("fetching fresh header from %q", p.host)
+	hdr, err := p.ec.HeaderByNumber(ctx, nil /* latest */)
+	if err != nil {
+		return nil, fmt.Errorf("HeaderByHash error: %w", err)
+	}
+	p.setTip(hdr)
+	return hdr, nil
+}
```
Comment (chappjc) [inline client/asset/eth/multirpc.go:150]: Error message should say `HeaderByNumber`.

But it looks like we also need something like the following to satisfy the `ethFetcher` interface:

```go
func (p *provider) headerByHash(ctx context.Context, hash common.Hash, log dex.Logger) (*types.Header, error) {
	log.Tracef("fetching block header for %v from %q", hash, p.host)
	hdr, err := p.ec.HeaderByHash(ctx, hash)
	if err != nil {
		return nil, fmt.Errorf("HeaderByHash error: %w", err)
	}
	return hdr, nil
}
```

**Review by chappjc — COMMENTED** [review]
Initial comments.  More today.

File: client/asset/eth/nodeclient_harness_test.go (hunk 16 above)
Comment (buck54321) [reply client/asset/eth/nodeclient_harness_test.go:124]: I do `TARGS="rpc tesnet" go test ...`, but yeah.

Comment (chappjc) [reply client/asset/eth/nodeclient_harness_test.go:124]: Flags work in tests, simpler:

```go
func TestMain(m *testing.M) {
	dexeth.MaybeReadSimnetAddrs()

	flag.BoolVar(&isTestnet, "testnet", false, "use testnet")
	flag.BoolVar(&useRPC, "rpc", false, "use RPC")
	flag.Parse()

	if isTestnet {
		if err := useTestnet(); err != nil {
			fmt.Fprintf(os.Stderr, "error loading testnet: %v", err)
			os.Exit(1)
		}
	}
```

test:

```go
func TestFlags(t *testing.T) {
	t.Logf("isTestnet? %v", isTestnet)
	t.Logf("useRPC? %v", useRPC)
}
```

```sh
$  go test -v -run TestFlags
=== RUN   TestFlags
    wsconn_test.go:40: isTestnet? false
    wsconn_test.go:41: useRPC? false
--- PASS: TestFlags (0.00s)
$  go test -v -run TestFlags -testnet
=== RUN   TestFlags
    wsconn_test.go:40: isTestnet? true
    wsconn_test.go:41: useRPC? false
--- PASS: TestFlags (0.00s)
$  go test -v -run TestFlags -testnet -rpc
=== RUN   TestFlags
    wsconn_test.go:40: isTestnet? true
    wsconn_test.go:41: useRPC? true
--- PASS: TestFlags (0.00s)
PASS
```

File: client/asset/eth/multirpc.go [hunk 19]
```diff
@@ -0,0 +1,1298 @@
...
+					log.Debugf("couldn't get a websocket connection for %q (original scheme: %q) (OK)", wsURL, ogScheme)
+				} else {
+					log.Errorf("failed to get websocket connection to %q. attempting http(s) fallback: error = %v", endpoint, err)
+				}
+			}
+		}
+		// Weren't able to get a websocket connection. Try HTTP now.
+		if ec == nil {
+			var err error
+			rpcClient, err = rpc.Dial(endpoint)
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/eth/multirpc.go:394]: Dial does path discrimination, and at this point, I'm no longer interested in introspection for the purposes of forcing a websocket connection. I'm just letting geth handle it now.

File: client/asset/eth/nodeclient_harness_test.go (hunk 16 above)
Comment (chappjc) [reply client/asset/eth/nodeclient_harness_test.go]: When and where you call Parse has gottas.  Maybe that was it

File: client/asset/eth/multirpc.go [hunk 20]
```diff
@@ -0,0 +1,1364 @@
...
+		confirmed:  confs > txConfsNeededToConfirm,
+	}
+	m.receipts.Unlock()
+
+	// TODO
+	// TODO: Plug into the monitoredTx system from #1638.
+	// TODO
+	if tx, _, err = m.getTransaction(ctx, txHash); err != nil {
+		return nil, nil, err
+	}
```
Comment (JoeGruffins) [inline client/asset/eth/multirpc.go:614]: Was this addition intentional? We already have the tx.

File: client/asset/eth/multirpc.go [hunk 21]
```diff
@@ -1109,35 +1118,46 @@ func (m *multiRPCClient) SubscribeFilterLogs(ctx context.Context, query ethereum
...
-	"mewapi.io",
-	"flashbots.net",
-	"mycryptoapi.com",
-	"runonflux.io",
-	"infura.io",
-	"rivet.cloud",
-	"alchemy.com",
+const (
+	// Compliant providers
+	providerIPC         = "IPC"
```
Comment (chappjc) [inline client/asset/eth/multirpc.go:1123]: I figured we would list this IPC one and include it in `compliantProviders` so the special case `p.host == ipcHost` check in `checkProvidersCompliance` could be removed.

Also we can use `providerIPC` in place of `ipcHost` like in `connectProviders`.

The last thing I way eying for these consts was using `providerRivetCloud` instead of the string literal in `connectProviders`.

File: client/asset/eth/multirpc.go [hunk 22]
```diff
@@ -0,0 +1,1364 @@
...
+	providerCloudflareETH: true,
+	providerAnkr:          true,
+}
+
+func providerIsCompliant(addr string) (known, compliant bool) {
+	addr = domain(addr)
+	if _, known = compliantProviders[addr]; known {
+		return true, true
+	}
+	_, known = compliantProviders[addr]
```
Comment (martonp) [inline client/asset/eth/multirpc.go:1155]: This should check `nonCompliantProviders`.

File: client/asset/eth/eth_test.go [hunk 23]
```diff
@@ -2989,21 +2990,21 @@ func TestDriverExists(t *testing.T) {
 	}
 
 	// Create the wallet.
-	err = CreateWallet(&asset.CreateWalletParams{
-		Type:     walletTypeGeth,
+	err = createWallet(&asset.CreateWalletParams{
```
Comment (chappjc) [inline client/asset/eth/eth_test.go:2993]: Apologies for the conflicts from putting @martonp's work in first.

```diff
diff --git a/client/asset/eth/eth_test.go b/client/asset/eth/eth_test.go
index 8b95ac68..ffac0eff 100644
--- a/client/asset/eth/eth_test.go
+++ b/client/asset/eth/eth_test.go
@@ -2986,7 +2986,7 @@ func TestDriverExists(t *testing.T) {
        settings := map[string]string{providersKey: "a.ipc"}
 
        // no wallet
-       exists, err := drv.Exists(walletTypeGeth, tmpDir, settings, dex.Simnet)
+       exists, err := drv.Exists(walletTypeRPC, tmpDir, settings, dex.Simnet)
        if err != nil {
                t.Fatalf("Exists error for no geth wallet: %v", err)
        }
@@ -3498,7 +3498,7 @@ func TestReconfigure(t *testing.T) {
        }
 
        walletCfg := &asset.WalletConfig{
-               Type:     walletTypeGeth,
+               Type:     walletTypeRPC,
                Settings: settings,
        }
 
@@ -4550,22 +4550,23 @@ func testMaxSwapRedeemLots(t *testing.T, assetID uint32) {
        logger := dex.StdOutLogger("ETHTEST", dex.LevelOff)
        tmpDir := t.TempDir()
 
-       err := CreateWallet(&asset.CreateWalletParams{
-               Type:     walletTypeGeth,
+       settings := map[string]string{providersKey: "a.ipc"}
+       err := createWallet(&asset.CreateWalletParams{
+               Type:     walletTypeRPC,
                Seed:     encode.RandomBytes(32),
                Pass:     encode.RandomBytes(32),
-               Settings: make(map[string]string),
+               Settings: settings,
                DataDir:  tmpDir,
                Net:      dex.Testnet,
                Logger:   logger,
-       })
+       }, true)
        if err != nil {
                t.Fatalf("CreateWallet error: %v", err)
        }
 
        wallet, err := drv.Open(&asset.WalletConfig{
-               Type:     walletTypeGeth,
-               Settings: make(map[string]string),
+               Type:     walletTypeRPC,
+               Settings: settings,
                DataDir:  tmpDir,
        }, logger, dex.Testnet)
        if err != nil {
```

File: client/asset/eth/multirpc.go [hunk 24]
```diff
@@ -0,0 +1,1368 @@
...
+		if time.Since(cached.lastAccess) < unconfirmedReceiptExpiration {
+			return cached.r
+		}
+	}
+	return nil
+}
+
+// cleanReceipts cleans up the receipt cache, deleting any receipts that haven't
+// been access for > receiptCacheExpiration.
+func (m *multiRPCClient) cleanReceipts() {
```
Comment (chappjc) [inline client/asset/eth/multirpc.go:564]: `cleanReceipts` isn't called anywhere.  Looks like you merged it into `cachedReceipt` and this method can be removed.

File: client/asset/eth/multirpc.go [hunk 25]
```diff
@@ -0,0 +1,1368 @@
...
+	parts := strings.Split(host, ".")
+	n := len(parts)
+	if n <= 2 {
+		return host
+	}
+	return parts[n-2] + "." + parts[n-1]
+}
+
+// checkProvidersCompliance verifies that a provider supports the API that DEX
+// requires by sending a series of requests and verfiying the responses. If a
```
Comment (chappjc) [inline client/asset/eth/multirpc.go:1314]: verfiying -> verifying

**Comment by martonp** [comment]
https://github.com/decred/dcrdex/pull/1832/files#r965796162

@buck54321 You fixed this in [325a552](https://github.com/decred/dcrdex/pull/1832/commits/325a552e4bf532b016b87c518d04ae43a5fc8bdb) but somehow it is back now.

**Review by martonp — APPROVED** [review]
